src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonValueOfT.cs

```diff
@@ -0,0 +1,375 @@
+
+        public JsonValue(TValue value, JsonNodeOptions? options = null) : base(options)
+        {
+            Debug.Assert(value != null);
+            Debug.Assert(!(value is JsonElement) || ((JsonElement)(object)value).ValueKind != JsonValueKind.Null);
+
+            if (value is JsonNode)
+            {
+                ThrowHelper.ThrowArgumentException_NodeValueNotAllowed(nameof(value));
+            }
+
+            _value = value;
+        }
+
+        public TValue Value
+        {
+            get
+            {
+                return _value;
+            }
+        }
+
+        public override T GetValue<[DynamicallyAccessedMembers(MembersAccessedOnRead)] T>()
+        {
+            Type returnType = typeof(T);
+
+            // If no conversion is needed, just return the raw value.
+            if (_value is T returnValue)
+            {
+                return returnValue;
+            }
+
+            if (_value is JsonElement jsonElement)
+            {
+                return ConvertJsonElement<T>();
+            }
+
+            // Currently we do not support other conversions.
+            // Generics (and also boxing) do not support standard cast operators say from 'long' to 'int',
+            //  so attempting to cast here would throw InvalidCastException.
+            throw new InvalidOperationException(SR.Format(SR.NodeUnableToConvert, _value!.GetType(), typeof(T)));
+        }
+
+        public override bool TryGetValue<[DynamicallyAccessedMembers(MembersAccessedOnRead)] T>([NotNullWhen(true)] out T value)
+        {
+            // If no conversion is needed, just return the raw value.
+            if (_value is T returnValue)
+            {
+                value = returnValue;
+                return true;
+            }
+
+            if (_value is JsonElement jsonElement)
+            {
+                return TryConvertJsonElement<T>(out value);
+            }
+
+            // Currently we do not support other conversions.
+            // Generics (and also boxing) do not support standard cast operators say from 'long' to 'int',
+            //  so attempting to cast here would throw InvalidCastException.
+            value = default!;
+            return false;
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2072:ystem.Text.Json.Node.JsonValue<TValue>.WriteTo(Utf8JsonWriter,JsonSerializerOptions): 'inputType' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicFields', 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Text.Json.JsonSerializer.Serialize(Utf8JsonWriter,Object,Type,JsonSerializerOptions)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.",
+            Justification = "The 'inputType' parameter if obtained by calling System.Object.GetType().")]
```

Comment:
Can we suppress this in our suppressions file, e.g. [this](https://github.com/dotnet/runtime/blob/c377f4152351dfbe16099317f553b87d2ce24b6b/src/libraries/System.Data.Common/src/ILLink/ILLink.Suppressions.xml#L70-L75)? A bit cleaner, I think.

---

Is it a high priority for usages of the node types to be ILLinker-friendly?

Reply:
I'm open to this, but don't think it's necessary with this PR.

Reply:
Considering for a future PR is fine with me.